rfcs/react-components/convergence/slot-children-render-function.md

```diff
@@ -0,0 +1,813 @@
+    Exclude<Slots[Key], SlotShorthandValue | (Key extends 'root' ? null : never)>
+  >;
+};
+```
+
+After:
+
+```ts
+type ComponentState<Slots extends SlotPropsRecord> = {
+  [Key in keyof Slots]: SlotComponent<ExtractSlotProps<Slots[Key]>>;
+};
+```
+
+##### Styles hooks `.props` access
+
+A minor modification in style hooks are required. at the moment we mutate the merged properties provided by `resolveShorthand`
+to include our custom `className` property.
+
+In this case we have some alternatives:
+
+###### Option 1 mutate external properties
+
+Instead of mutating the merged properties, simply mutate the External properties from the slot component:
+
+```diff
+// useAccordionHeader
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
+  const styles = useStyles();
+- state.root.className = mergeClasses(
++ state.root.props.className = mergeClasses(
+    accordionHeaderClassNames.root,
+    styles.root,
+    state.inline && styles.rootInline,
+    state.disabled && styles.rootDisabled,
+-   state.root.className,
++   state.root.props.className,
+  );
+
+  if (state.icon) {
+-    state.icon.className = mergeClasses(
++    state.icon.props.className = mergeClasses(
+      accordionHeaderClassNames.icon,
+      styles.icon,
+-      state.icon.className
++      state.icon.props.className
+    );
+  }
+
+  return state;
+};
+```
+
+###### Option 2 mutate overrides
+
+By introducing an `overrides` property on the `ComponentState` we can create a layer of override that can be mutated:
+
+```diff
+// useAccordionHeader
+export const useAccordionHeaderStyles_unstable = (state: AccordionHeaderState) => {
+  const styles = useStyles();
+- state.root.className = mergeClasses(
++ state.overrides.root.className = mergeClasses(
+    accordionHeaderClassNames.root,
+    styles.root,
+    state.inline && styles.rootInline,
+    state.disabled && styles.rootDisabled,
+-   state.root.className,
++   state.root.props.className,
+  );
+
+    if (state.icon) {
+-    state.icon.className = mergeClasses(
++    state.overrides.icon.className = mergeClasses(
+      accordionHeaderClassNames.icon,
+      styles.icon,
+-      state.icon.className
++      state.icon.props.className
+    );
+  }
+
+  return state;
+};
+```
+
+On the render function side, we just need to spread the overrides:
+
+```tsx
+export const renderAccordionHeader_unstable = (state: AccordionHeaderState) => (
+  <state.root {...state.overrides.root}>
+    <state.button>
+      {state.expandIconPosition === 'start' && state.expandIcon && <state.expandIcon />}
+      {state.icon && <state.icon />}
+      {state.root.props.children}
+      {state.expandIconPosition === 'end' && state.expandIcon && <state.expandIcon />}
+    </state.button>
+  </state.root>
+);
+```
+
+###### Option 3 stop mutating
+
+Let's just stop mutating, and since we require a layer of custom styling through classNames
+let's just properly use an argument on the render method exclusively for this:
```

Comment:
But to be clear, the option 3 was focused solely on creating a new signature for the `styles` only, not for all overrides

Reply:
> The other way to merge the props together would be having a `state.overrides.SLOT` , and then simply adding the overrides on render function.

Like I mentioned in my other comment https://github.com/microsoft/fluentui/pull/27164#discussion_r1149605391, I don't think `state.overrides.SLOT` will end up being a good idea: it gets very tricky to check whether you need to merge with `overrides` vs. `props` vs. `defaultProps`. Instead, I think mutating `state.SLOT.props` is the best option for overrides.

> But to be clear, the option 3 was focused solely on creating a new signature for the `styles` only, not for all overrides

The change to the `useComponentStyles` hooks seems like a separate RFC from this one. It might be better to split that proposal out, in order to keep this one focused on the other changes (JSX pragma, etc.)